internals/controllers/migrationhierarchy/migrationhierarchy_controller.go

```diff
@@ -0,0 +1,440 @@
+}
+
+func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext) error {
+	ctx := mhObject.Ctx
+	logger := log.FromContext(ctx)
+
+	currentNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.CurrentNamespace
+	toNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.ToNamespace
+
+	// get the namespace of the subnamespace that should be migrated
+	ns, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: currentNamespace}, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("failed getting namespace object '%s': "+err.Error(), currentNamespace)
+	}
+
+	sourceSNSParentName := utils.GetNamespaceParent(ns.Object)
+	oldSNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: currentNamespace, Namespace: sourceSNSParentName}, &danav1.Subnamespace{})
+	if err != nil {
+		return fmt.Errorf("failed getting subnamespace object '%s': "+err.Error(), sourceSNSParentName)
+	}
+
+	// create a new subnamespace under the new parent
+	newSNS, reason, er := r.createSNS(oldSNS, currentNamespace, toNamespace)
+	if er != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed creating subnamespace '%s' under namespace '%s': "+er.Error(), currentNamespace, toNamespace)
+	}
+	logger.Info("successfully created new subnamespace under new parent", "subnamespace", currentNamespace, "new parent", toNamespace)
+
+	// delete the old subnamespace under the original parent
+	if reason, er := r.deleteSNS(oldSNS); er != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+
+		// if deleting the old subnamespace fails, delete the newly created subnamespace
+		if reason, er := r.deleteSNS(newSNS); er != nil {
+			err := r.updateMHErrorStatus(mhObject, er, reason)
+			if err != nil {
+				return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+			}
+			return fmt.Errorf("failed deleting subnamespace '%s' :"+err.Error(), newSNS.GetName())
+		}
+		return fmt.Errorf("failed deleting subnamespace '%s' :"+er.Error(), oldSNS.GetName())
+	}
+	logger.Info("successfully deleted old subnamespace from old parent", "subnamesapce", currentNamespace, "old parent", sourceSNSParentName)
+
+	toNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: toNamespace}, &corev1.Namespace{})
+	if err != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), toNamespace)
+	}
+
+	if err := r.UpdateNSBasedOnParent(ctx, toNS, ns); err != nil {
```

Comment:
ill make a function handle the related objects updates

Reply:
Wrapped